@here/harp-mapview/lib/VisibleTileSet.ts

```diff
@@ -831,143 +671,64 @@ export class VisibleTileSet {
-
-                if (area === undefined) {
-                    throw new Error("Unexpected tile key");
-                }
-
-                if (area <= 0 || tileKey.level > displayZoomLevel) {
-                    continue;
-                }
-
-                if (dataSource.shouldRender(displayZoomLevel, tileKey)) {
-                    visibleTiles.push(tileEntry);
-                }
-
-                tilingScheme.getSubTileKeys(tileKey).forEach(childTileKey => {
-                    const offset = tileEntry.offset;
-                    const tileKeyAndOffset = TileOffsetUtils.getKeyForTileKeyAndOffset(
-                        childTileKey,
-                        offset
-                    );
-                    const intersectsFrustum = tileFrustumIntersectionCache.get(tileKeyAndOffset);
-
-                    if (intersectsFrustum !== undefined) {
-                        return;
-                    }
-
-                    const geoBox = this.getGeoBox(tilingScheme, childTileKey, offset);
-
-                    if (useElevationRangeSource) {
-                        const range = elevationRangeSource!.getElevationRange(childTileKey);
-                        geoBox.southWest.altitude = range.minElevation;
-                        geoBox.northEast.altitude = range.maxElevation;
-
-                        allBoundingBoxesFinal =
-                            allBoundingBoxesFinal &&
-                            range.calculationStatus === CalculationStatus.FinalPrecise;
-                    }
-
-                    let subTileArea = 0;
-
-                    if (this.projection.type === ProjectionType.Spherical) {
-                        const obb = new OrientedBox3();
-                        this.options.projection.projectBox(geoBox, obb);
-                        if (obb.intersects(this.m_frustum)) {
-                            subTileArea = 1;
-                        }
-                    } else {
-                        this.options.projection.projectBox(geoBox, tileBounds);
-                        subTileArea = this.computeSubTileArea(tileBounds);
-                    }
-
-                    tileFrustumIntersectionCache.set(tileKeyAndOffset, subTileArea);
+            const bucket = dataSourceBuckets.get(tilingScheme);
+            if (bucket === undefined) {
+                dataSourceBuckets.set(tilingScheme, [dataSource]);
+            } else {
+                bucket.push(dataSource);
+            }
+        });
 
-                    if (subTileArea > 0) {
-                        workList.push(new TileKeyEntry(childTileKey, subTileArea, offset));
+        this.m_frustumIntersection.updateFrustum();
+
+        // For each bucket of data sources with same tiling scheme, calculate frustum intersection
+        // once using the maximum display level.
+        for (const [tilingScheme, bucket] of dataSourceBuckets) {
+            const maxDisplayLevel = Math.max(
```

Comment:
This looks like we could generate tiles which the DataSource has no data for, say for example you have in the map call below, 10 & 12, then maxDisplayLevel would be 12, and it would generate tileKeyEntries at 12, however the datasource which has max zoom level 10 would not show any data(?)

I guess then you need to differentiate the DataSources based on the max zoom levels? Or did I miss something?

Reply:
We check this in the next loop where we call dataSource.shouldRender(), which returns false if the tile belongs to a level higher than the data source maximum level.